### PR TITLE
ARROW-17081: [Java][Datasets] Move JNI build configuration from cpp/ to java/

### PR DESCRIPTION
**Pull request by kou**

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
@lwhite1 @davisusanibar This is ready to review/merge.
(But this is difficult to review without CMake skill...)

I think that "Java / AMD64 Windows Server 2022 Java JDK 11" failure is unrelated: https://github.com/apache/arrow/runs/8122792610?check_suite_focus=true

Could you confirm it?

**Comment by kou** [comment]
@github-actions crossbow submit java-jars

**Comment by kou** [comment]
+1

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -851,6 +864,9 @@ add_dependencies(arrow_test_dependencies toolchain-tests)
 if(ARROW_STATIC_LINK_LIBS)
   add_dependencies(arrow_dependencies ${ARROW_STATIC_LINK_LIBS})
   if(ARROW_HDFS OR ARROW_ORC)
+    if(Protobuf_SOURCE STREQUAL "SYSTEM")
```
Comment (pitrou) [inline cpp/CMakeLists.txt:867]: @kou This seems to require protobuf when HDFS is enabled, is there any reason for it? Is it a mistake?

Comment (kou) [reply cpp/CMakeLists.txt:867]: Oh, good catch! It might be a mistake. Let's try removing this: #39136

**Pull request merged**